File: meetmind/database/rag_retriever.py
"""每个 Agent 私有的 RAG 检索工具：ES 混合检索 + Cohere rerank。

流程（针对一次 query）：
  1) BM25 检索：ES match query 拉 top_n_bm25 候选；
  2) 向量检索：query → embedding，ES dense_vector kNN 拉 top_n_knn 候选；
  3) 合并去重：按 ES _id 取并集；
  4) Cohere Rerank：把合并后的候选送 `rerank-v4.0-pro`，按相关性重排；
  5) 取 rerank 后的 top-K 给 LLM。

把检索器通过 `to_tool()` 暴露成 LangChain Tool，LLM 用 function-calling
按需调用，由 `BaseAgent.invoke()` 的 tool_calls 循环执行。
"""

from __future__ import annotations

from dataclasses import dataclass

from langchain_core.tools import Tool

from meetmind.config.settings import get_settings
from meetmind.database.client import get_agent_index, get_es_client
from meetmind.database.embedding import embed
from meetmind.database.reranker import rerank, RerankedDoc
from meetmind.utils.logger import get_logger

logger = get_logger(__name__)


# ---------------------------------------------------------------- 数据类
@dataclass
class RetrievedDoc:
    """混合检索 + rerank 之后命中的单条结果。

    - `content`         原始文档文本
    - `metadata`        灌库时附加的 type / date / source
    - `relevance_score` Cohere rerank 的相关性分数 (0~1, 越大越相关)
    """

    content: str
    metadata: dict
    relevance_score: float = 0.0

    def as_context_line(self) -> str:
        """格式化为单行字符串，便于拼成 prompt 上下文块。"""
        meta = self.metadata or {}
        tag = meta.get("type") or "note"
        date = meta.get("date") or ""
        if date:
            return f"- [{tag} / {date}] {self.content}"
        return f"- [{tag}] {self.content}"



class RAGRetriever:
    """绑定到单个 agent 的 ES index，提供 (BM25 + 向量) 混合检索 + Cohere rerank。

    每个 Agent 实例化时持有一个独立的 RAGRetriever。
    `restart()` 在每轮 Agent.invoke() 开头被调用，只用于清零 call_count。
    """

    def __init__(self, agent_name: str):
        self.agent_name = agent_name
        # ensure_agent_index 是幂等的；这里调一次保证 index 存在
        self.index_name = get_agent_index(agent_name)
        # 本轮调用计数，BaseAgent 通过它判断是否要在面板上标 "📚 调用过 RAG"
        self.call_count: int = 0

    def restart(self) -> None:
        """每轮 Agent.invoke() 开始前调用，清零调用次数。"""
        self.call_count = 0

    # client.search()返回的结果的结构:
    # {
    #     "hits": {
    #         "total": {"value": 100},
    #         "hits": [
    #             {
    #                 "_id": "abc123",
    #                 "_score": 1.5,
    #                 "_source": {
    #                     "content": "用户认证采用 JWT 方案",
    #                     "metadata": {"type": "note", "date": "2024-01"}
    #                 }
    #             },
    #             {
    #                 同上
    #             }
    #         ]
    #     }
    # }

    def _bm25_search(self, query: str, size: int) -> list[dict]:
        """ES BM25 关键词检索。返回 list of {_id, score, _source} """
        client = get_es_client()

        # body 是 ES的 向量检索 指定请求格式
        # body 结构参考 ES DSL：https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html
        body = {
            "query": {
                "match": {
                    "content": query
                }
            },
            "_source": ["content", "metadata"],
            "size": size,
        }
        try:
            resp = client.search(index=self.index_name, body=body)
        except Exception as exc:
            logger.warning(f"[{self.agent_name}] BM25 检索失败: {exc}")
            return []
        return list(resp["hits"]["hits"])

    def _knn_search(self, query: str, size: int) -> list[dict]:
        """ES dense_vector kNN 向量检索。"""
        client = get_es_client()
        query_vec = embed(query)
        # body 是 ES的 knn 指定请求格式
        # body 结构参考 ES DSL：https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-knn-query.html

        body = {
            "knn": {
                "field": "embedding",
                "query_vector": query_vec,
                "k": size,
                "num_candidates": max(size * 5, 100),
            },
            "_source": ["content", "metadata"],
            "size": size,
        }
        try:
            resp = client.search(index=self.index_name, body=body)
        except Exception as exc:
            logger.warning(f"[{self.agent_name}] kNN 检索失败: {exc}")
            return []
        return list(resp["hits"]["hits"])

    @staticmethod
    def _merge(bm25_hits: list[dict], knn_hits: list[dict]) -> list[dict]:
        """按 _id 取并集去重；保持顺序（BM25 在前，kNN 补充）。"""
        seen: set[str] = set()
        merged_content: list[dict] = []
        for hit in bm25_hits:
            doc_id = hit["_id"]
            if doc_id in seen:
                continue
            seen.add(doc_id)
            merged_content.append(hit)
        for hit in knn_hits:
            doc_id = hit["_id"]
            if doc_id in seen:
                continue
            seen.add(doc_id)
            merged_content.append(hit)
        return merged_content

    # ---------- 公开检索接口 ----------
    def retrieve(self, query: str, top_n: int | None = None) -> list[RetrievedDoc]:
        """混合检索 + Cohere rerank，返回最终的 top_n 条 RetrievedDoc。

        参数 `top_n` 是 rerank 之后给 LLM 的条数，默认取 Settings.rerank_top_n（5）。
        BM25 / kNN 各自捞的候选数走 Settings.retrieve_top_n（20）。
        """

        # 拿到 召回数 和 rerank 数
        settings = get_settings()
        # 调用一次 检索
        self.call_count += 1

        if top_n is None:
            top_n = settings.rerank_top_n
        candidate_n = settings.retrieve_top_n

        # 1) 两路检索
        knn_hits = self._knn_search(query, size=candidate_n)
        bm25_hits = self._bm25_search(query, size=candidate_n)

        # 2) 合并去重
        merged_hits = self._merge(bm25_hits, knn_hits)
        if not merged_hits:
            logger.info(f"[{self.agent_name}] 混合检索: 命中 0 条")
            return []

        logger.info(
            f"[{self.agent_name}] 混合检索: BM25 检索到{len(bm25_hits)}个 chunk + kNN 检索到{len(knn_hits)}个chunk "
            f"→ 去重后 {len(merged_hits)} 条chunks 送入 rerank"
        )

        # 3) Cohere rerank
        source_contents: list[str] = []
        for hit in merged_hits:
            # 拿到 文本数据
            content = hit["_source"].get("content", "")
            source_contents.append(content)

        reranked : list[RerankedDoc] = rerank(query=query, documents=source_contents, top_n=top_n)

        # 4) 按 rerank 顺序拼 RetrievedDoc
        result: list[RetrievedDoc] = []
        for r in reranked:
            hit = merged_hits[r.index]
            source = hit["_source"]
            doc = RetrievedDoc(
                content=source.get("content", ""),
                metadata=source.get("metadata") or {},
                relevance_score=r.relevance_score,
            )
            result.append(doc)

        return result



    def get_tool(self) -> Tool:
        """
        把_rag_search 封装成 LangChain Tool，供 LLM function-calling 调用。
        """

        def _rag_search(query: str) -> str:
            """返回用于 prompt / 工具结果的格式化上下文块。"""

            # 用 retrieve 获取文档
            docs = self.retrieve(query)
            if not docs:
                return "(知识库中未找到相关条目)"
            lines: list[str] = []
            for d in docs:
                lines.append(d.as_context_line())
            return "\n".join(lines)

        description : str= (
            f"检索 {self.agent_name} 的私有知识库（工作日志 / 代码片段 / 设计文档）。"
            "当你认为当前问题可能与历史经验、既有约定、过往实现相关时调用本工具；"
            "若问题与本角色的历史经验无关，可不调用。"
            "参数 query：自然语言查询字符串。"
        )
        return Tool(name=f"rag_search_{self.agent_name}", description=description, func=_rag_search)

File: src/database/rag_retriever.ts
/**
 * 每个 Agent 私有的 RAG 检索工具：PostgreSQL 混合检索 + 本地 cross-encoder rerank。
 *
 * 流程：
 *   1) 关键字召回：pg_trgm word_similarity(query, content) 拉 candidateN 候选
 *   2) 向量检索：query → embedding，pgvector cosine kNN 拉 candidateN 候选
 *   3) 合并去重：按行 id 取并集
 *   4) 本地 cross-encoder rerank
 *   5) 取 rerank 后的 top-K 给 LLM
 *
 * 检索逻辑（混合检索 + rerank）在 `retrieve()` 里；把它包成 LangChain Tool 的工厂
 * 见 `src/tools/ragSearchTool.ts`。
 */

import { getSettings } from "../config/settings.js";
import { getLogger } from "../utils/logger.js";
import { ensureAgentTable, getPgPool } from "./client.js";
import { embed } from "./embedding.js";
import { rerank } from "./reranker.js";

const logger = getLogger("database.rag_retriever");

/**
 * 混合检索 + rerank 之后命中的单条结果。
 */
export interface RetrievedDoc {
  content: string;
  metadata: Record<string, unknown>;
  relevanceScore: number;
}

interface DbRow {
  id: string;
  content: string;
  metadata: Record<string, unknown> | null;
  score: number;
}

export class RAGRetriever {
  readonly agentName: string;
  // 本轮调用计数；BaseAgent 通过它判断是否要在面板上标 "调用过 RAG"
  callCount: number = 0;
  private _tableName: string | null = null;

  /** 绑定该检索器到指定 agent，表名由 getTableName(agentName) 决定。 */
  constructor(agentName: string) {
    this.agentName = agentName;
  }

  /**
   * 每轮 Agent.invoke() 开始前调用，清零调用次数。
   */
  restart(): void {
    this.callCount = 0;
  }

  /** 懒加载本 agent 对应的表名；首次会触发建表（含扩展 / 索引）。 */
  private async getTableName(): Promise<string> {
    if (this._tableName === null) {
      this._tableName = await ensureAgentTable(this.agentName);
    }
    return this._tableName;
  }

  /**
   * 关键字召回：pg_trgm 的 word_similarity 衡量「query 与 content 某片段」的 trigram 相似度。
   * 只取有任何重叠（>0）的候选，按相似度降序拉 size 条。中文召回偏粗，但后面有 rerank 兜底。
   */
  private async bm25Search(query: string, size: number): Promise<DbRow[]> {
    const pool = getPgPool();
    try {
      const tableName = await this.getTableName();
      const sql =
        `SELECT id, content, metadata, word_similarity($1, content) AS score ` +
        `FROM ${tableName} ` +
        `WHERE word_similarity($1, content) > 0 ` +
        `ORDER BY score DESC ` +
        `LIMIT $2`;
      const resp = await pool.query(sql, [query, size]);
      return resp.rows as DbRow[];
    } catch (exc) {
      logger.warning(`[${this.agentName}] 关键字检索失败: ${String(exc)}`);
      return [];
    }
  }

  /**
   * 向量检索：query 算 embedding 后用 pgvector cosine 距离算子 `<=>` 取最近的 size 条。
   * score = 1 - 距离，语义上等价于余弦相似度（越大越相关）。
   */
  private async knnSearch(query: string, size: number): Promise<DbRow[]> {
    const pool = getPgPool();
    try {
      const tableName = await this.getTableName();
      const queryVec = await embed(query);
      const vectorLiteral = `[${queryVec.join(",")}]`;
      const sql =
        `SELECT id, content, metadata, 1 - (embedding <=> $1::vector) AS score ` +
        `FROM ${tableName} ` +
        `ORDER BY embedding <=> $1::vector ` +
        `LIMIT $2`;
      const resp = await pool.query(sql, [vectorLiteral, size]);
      return resp.rows as DbRow[];
    } catch (exc) {
      logger.warning(`[${this.agentName}] kNN 检索失败: ${String(exc)}`);
      return [];
    }
  }

  /** 关键字与向量两路命中按行 id 取并集；关键字结果优先排前，向量结果补在后。 */
  private merge(bm25Hits: DbRow[], knnHits: DbRow[]): DbRow[] {
    const seen = new Set<string>();
    const merged: DbRow[] = [];
    for (const hit of bm25Hits) {
      if (seen.has(hit.id)) {
        continue;
      }
      seen.add(hit.id);
      merged.push(hit);
    }
    for (const hit of knnHits) {
      if (seen.has(hit.id)) {
        continue;
      }
      seen.add(hit.id);
      merged.push(hit);
    }
    return merged;
  }

  /**
   * 混合检索 + 本地 rerank。`topN` 是 rerank 之后给 LLM 的条数。
   */
  async retrieve(query: string, topN?: number): Promise<RetrievedDoc[]> {
    this.callCount += 1; // 记一次调用，供面板标「用过 RAG」

    const settings = getSettings();
    const finalN = topN ?? settings.rerankTopN;
    const candidateN = settings.retrieveTopN;

    // 1) 关键字与向量两路并行检索，互不阻塞
    const [knnHits, bm25Hits] = await Promise.all([
      this.knnSearch(query, candidateN),
      this.bm25Search(query, candidateN),
    ]);

    // 2) 合并去重
    const merged = this.merge(bm25Hits, knnHits);
    if (merged.length === 0) {
      logger.info(`[${this.agentName}] 混合检索: 命中 0 条`);
      return [];
    }

    logger.info(
      `[${this.agentName}] 混合检索: 关键字 ${bm25Hits.length} + kNN ${knnHits.length} → 去重后 ${merged.length} 条 送入 rerank`,
    );

    // 3) 本地 cross-encoder rerank：取每条候选正文送重排
    const sourceContents: string[] = [];
    for (const hit of merged) {
      sourceContents.push(hit.content ?? "");
    }
    const reranked = await rerank(query, sourceContents, finalN);

    // 4) 按 rerank 顺序拼 RetrievedDoc
    const result: RetrievedDoc[] = [];
    for (const r of reranked) {
      const hit = merged[r.index];
      result.push({
        content: hit.content ?? "",
        metadata: hit.metadata ?? {},
        relevanceScore: r.relevanceScore,
      });
    }
    return result;
  }
}

// 每个 agent 一个 RAGRetriever，按 agent 名缓存复用。
// BaseAgent 和 rag_search 工具都通过 getRetriever 拿同一个实例，
// 这样工具里的检索调用次数能被 agent 的 used_rag 统计（callCount）感知到。
const _retrievers = new Map<string, RAGRetriever>();

/** 取（必要时新建）某 agent 的 RAGRetriever 单例。 */
export function getRetriever(agentName: string): RAGRetriever {
  let retriever = _retrievers.get(agentName);
  if (retriever === undefined) {
    retriever = new RAGRetriever(agentName);
    _retrievers.set(agentName, retriever);
  }
  return retriever;
}
